src/zimdump.cpp

```diff
@@ -253,21 +271,47 @@ void ZimDumper::dumpFiles(const std::string& directory)
   ::mkdir(directory.c_str(), 0777);
 #endif
 
-  std::set<char> ns;
+  std::vector<std::string> pathcache;
+  std::set<char> nscache;
   for (zim::File::const_iterator it = pos; it != file.end(); ++it)
   {
-    std::string d = directory + '/' + it->getNamespace();
-    if (ns.find(it->getNamespace()) == ns.end())
-#if defined(_WIN32)
+    char filenamespace = it->getNamespace();
+    std::string d = directory + '/' + filenamespace;
+    if (nscache.find(filenamespace) == nscache.end())
+    {
+    #if defined(_WIN32)
       ::mkdir(d.c_str());
-#else
+    #else
       ::mkdir(d.c_str(), 0777);
-#endif
+    #endif
+        nscache.insert(filenamespace);
+    }
+    std::string t = it->getTitle();
     std::string url = it->getUrl();
-    std::string::size_type p;
-    while ((p = url.find('/')) != std::string::npos)
-      url.replace(p, 1, "%2f");
-    if ( url.length() > 255 )
+
+    if (filenamespace == 'A')
+    {
+        std::string::size_type p;
+        while ((p = url.find('/')) != std::string::npos)
+            url.replace(p, 1, "%2f");
+    }
+    else
+    {
+        // non-A namespaces may have nested directory structures
+        auto position = t.find_last_of('/');
+        if (position != std::string::npos)
+        {
+            std::string path = t.substr(0, position);
+            if (find(pathcache.begin(), pathcache.end(), path) == pathcache.end())
+            {
+                std::string base = directory + '/' + filenamespace + '/';
+                createdir(t, base);
+                pathcache.push_back(path);
+            }
+        }
+    }
```

Comment:
@mgautierfr the issue here is that the filenames contain valid slashes "/". So it is impossible to know if the '/' is a directory or if its just part of the name.

Reply:
@MiguelRocha I see no reason why the article with URL `AC/DC` shoud not be written in the path `AC/DC`: `AC` being a directory and `DC` a file with the HTML. By doing the `/` encoding you are breaking the HTML rendering, see http://library.kiwix.org/wikipedia_en_all_nopic/A/AC/DC

Reply:
A "/" is used as a separator when serving the content through a web server.
The links are relative and take into account "/" in the path (the target of the link and the "position" of the source article").

This is why we do html redirection instead of symlink when we dump articles.
This is to ensure that the relative link is correct base on the source article.

We must follow the directory structure exposed as article path to ensure that a static file server correctly find articles.

Reply:
> @MiguelRocha I see no reason why the article with URL `AC/DC` shoud not be written in the path `AC/DC`: `AC` being a directory and `DC` a file with the HTML. By doing the `/` encoding you are breaking the HTML rendering, see http://library.kiwix.org/wikipedia_en_all_nopic/A/AC/DC

This was causing me some confusion but now its clear.

@mgautierfr thanks for the explanation. I will do this change.

Reply:
I'm having an issue now where the redirect articles have the same name as some sub-directories. That makes it impossible to create the sub-directories in 'A' namespace. There may be an issue with the way we write redirect articles.